### ISSUE TITLE
Feature Proposal: Trigger password selection dialog with url scheme action

### ISSUE THREAD

**Issue by torstenlehmann** [MacPass/MacPass#1066]
**First:** Thanks for this awesome project! It's almost exactly what I need and it's so well made that I'm motivated to contribute.

**Is your feature request related to a problem? Please describe.**
- When I global autotype the selection process is not under my full control, because I can't always predict which item will be matched.
- When I switch to MacPass and select the exact item to autotype I loose my current context.
- Matching on the window title is often not sufficient, especially if I'm not able to replace the title or extend it with e.g. a url.

**Describe the solution you'd like**
For a proposal see this message

**Describe alternatives you've considered**
We can use AppleScript (https://github.com/georgesnow/MPSafariAF) to switch back and forth from the current app into MacPass. This is a solution, but we loose the current context again.

## Proposal

**Step 1:** Open login web page (e.g. `github.com`)

<img width="330" alt="github_login" src="https://user-images.githubusercontent.com/5949960/78506808-889f1b00-777c-11ea-8985-791d05150bf1.png">

**Step 2:** Call url action (via Browser extension, AppleScript, Alfred Workflow)

`macpass://selectPassword?title={windowTitle}&url={websiteURLOrOtherLocation}&app={bundleIdentifierOrAppName}`

e.g. `macpass://selectPassword?title=Sign%20in%20to%20GitHub%20%C2%B7%20GitHub&url=https%3A%2F%2Fgithub.com%2Flogin`

- We could also make it configurable if we ask for username, password or both (regarding https://github.com/MacPass/MacPass/issues/1039)
- The exact API and matching rules need further discussion.

**Result:** Selection dialog appears

- The usage of url schemes leads to a MacPass launch **on demand**!
- MacPass main window is hidden, no MacPass menu is visible
- It may be configurable if we need to see the dialog if only one item matches (so we get the current behaviour, but I prefer to see it always, knowing exactly what is typed in).

**Step 3:** Unlock database in place (if needed)

- If database is locked, it provides the ability to unlock in place
- UX is just a prototype

<img width="572" alt="unlock_database" src="https://user-images.githubusercontent.com/5949960/78506821-a9677080-777c-11ea-8275-a13b3d8c5219.png">

**Step 4:** Select item

<img width="572" alt="select_item" src="https://user-images.githubusercontent.com/5949960/78506827-aec4bb00-777c-11ea-9431-547dffb20244.png">

- UX is just a prototype
- You can also open MacPass from here if your item of interest is not in this list
- ESC closes the window (like spotlight, alfred)

**Bottom line:**

With an API like url schemes we can streamline the selection process because MacPass does not need to pull specific data from apps (like window title), but it gets more rich data **pushed** from the app, tool, browser extension itself. MacPass then launches on demand, has more context to match items (username, password or both) and can show a more streamlined GUI without context switches. But I see the point in having the pull based flow too, for people not having tools that can call API url schemes, but Browser extensions should be simple to write this way, if the permission model allows it.

**Comment by torstenlehmann** [comment]
So thinking further about it, maybe we should split it up into:
- Launching MacPass on demand with action to execute
- Getting/providing more context than window title (maybe Plugins are sufficient?)
- Matching rules that differentiate between title, URL, application or combination?
- More minimal selection GUI
- Unlock database in selection GUI on demand
- Option to open MacPass main window from selection GUI
- Option to always show selection GUI, if I pressed global autofill

**Comment by georgesnow** [comment]
Does the AutotypeURL plugin sort of fit this need? That script was folded into that plugin. There is also Macpass Revealer plugin but thats in process to be folded into the core code of MacPass. There is branch that incorporates that safari AppleScript. If Safari is the front most window. And you trigger a user configured hotkey. It triggers that AppleScript (part of it) and shows macpass sets the search contest based on that URL.


Or am I misunderstanding what you hoping to achieve? It sounds like you are also looking for URI scheme like 1Password.

**Comment by torstenlehmann** [comment]
> Does the AutotypeURL plugin sort of fit this need?
It's not exactly what I thought of, but it solves my Safari integration problem for now, as long I use a title like "subdomain.domain.com (accountName)", because it only seems to take the external url as title and matches against a database entry's title.

MacPass Revealer's functionality ist not completely clear to me. But it's nice if MacPass can run in a status item in the future.

I have an issue that everything needs to be triggered with a global shortcut. I don't like to polute my global shortcut space with things that I can trigger via Alfred. So I would need a different interface to trigger MacPass actions instead of global shortcuts.
